crates/data_components/src/mssql.rs

```diff
@@ -0,0 +1,619 @@
+    }
+
+    async fn query_arrow(
+        &self,
+        sql: &str,
+        projected_schema: SchemaRef,
+    ) -> Result<Vec<RecordBatch>, Box<dyn std::error::Error + Send + Sync>> {
+        tracing::debug!("Executing sql: {sql}");
+
+        let schema = Arc::clone(&projected_schema);
+
+        let mut conn = self.conn.get().await.boxed().context(ConnectionPoolSnafu)?;
+
+        let query_res = conn
+            .simple_query(sql)
+            .await
+            .boxed()
+            .context(ConnectionPoolSnafu)?
+            .into_row_stream();
+
+        let mut chunked_stream = query_res.chunks(4_000).boxed();
+
+        let mut records: Vec<RecordBatch> = Vec::new();
+
+        while let Some(chunk) = chunked_stream.next().await {
+            let rows = chunk
+                .into_iter()
+                .collect::<Result<Vec<_>, _>>()
+                .context(QuerySnafu)
+                .map_err(to_datafusion_err)?;
+
+            records.push(rows_to_arrow(&rows, &schema)?);
+        }
+
+        Ok(records)
+    }
+}
+
+macro_rules! handle_primitive_type {
+    ($builder:expr, $type:expr, $builder_ty:ty, $value_ty:ty, $row:expr, $index:expr) => {{
+        let Some(builder) = $builder else {
+            return NoBuilderForIndexSnafu { index: $index }.fail();
+        };
+        let Some(builder) = builder.as_any_mut().downcast_mut::<$builder_ty>() else {
+            return FailedToDowncastBuilderSnafu {
+                mssql_type: format!("{:?}", $type),
+            }
+            .fail();
+        };
+        let v = $row.get::<$value_ty, usize>($index);
+        match v {
+            Some(v) => builder.append_value(v),
+            None => builder.append_null(),
+        }
+    }};
+}
+
+#[allow(clippy::too_many_lines)]
+fn rows_to_arrow(rows: &[Row], schema: &SchemaRef) -> Result<RecordBatch> {
+    let mut arrow_columns_builders: Vec<Option<Box<dyn ArrayBuilder>>> = Vec::new();
+    let mut mssql_types: Vec<ColumnType> = Vec::new();
+    let mut column_names: Vec<String> = Vec::new();
+
+    if !rows.is_empty() {
+        let row = &rows[0];
+        for column in row.columns() {
+            let column_name = column.name();
+            let column_type = column.column_type();
+            let (decimal_precision, decimal_scale) = match column_type {
+                ColumnType::Decimaln | ColumnType::Numericn => {
+                    // use 38, 10 as default precision and scale for decimal types
+                    let (precision, scale) =
+                        get_column_precision_and_scale(column_name, schema).unwrap_or((38, 10));
+                    (Some(precision), Some(scale))
+                }
+                _ => (None, None),
+            };
+            let data_type =
+                map_column_type_to_arrow_type(column_type, decimal_precision, decimal_scale);
+
+            arrow_columns_builders.push(map_data_type_to_array_builder_optional(Some(&data_type)));
```

Comment:
```suggestion
            arrow_columns_builders.push(map_data_type_to_array_builder(&data_type));
```
Can't we use the non-optional one here?

Reply:
Missed that we have non-optional one - updated. Thanks